code/parse/sexp/sexp_lookup.cpp

```diff
@@ -237,31 +218,29 @@ void dynamic_sexp_shutdown()
 	auto& global = globals();
 
 	global.operator_const_mapping.clear();
-	global.next_free_operator_mapping.clear();
+	global.subcategory_to_category.clear();
+
+	global.next_free_operator_id = First_available_operator_id;
+	global.next_free_category_id = First_available_category_id;
+	global.next_free_subcategory_id = First_available_subcategory_id;
 
 	global.initialized = false;
 }
-int add_subcategory(int parent_category, const SCP_string& name)
+int add_category(const SCP_string& name)
 {
-	// Another hack to make sure change2 is interpreted as the normal change category
-	if (parent_category == OP_CATEGORY_CHANGE2) {
-		parent_category = OP_CATEGORY_CHANGE;
-	}
-
-	int max_subcategory_value = 0;
-	for (auto& subcat : op_submenu) {
-		if ((subcat.id & OP_CATEGORY_MASK) == parent_category) {
-			max_subcategory_value = std::max(max_subcategory_value, subcat.id & SUBCATEGORY_MASK);
-		}
-	}
+	auto& global = globals();
 
-	if (max_subcategory_value >= SUBCATEGORY_MASK) {
-		// No more entries left in this subcategory!
-		return -1;
-	}
+	int category_id = global.next_free_category_id++;
+	op_menu.push_back({ name, category_id });
+	return category_id;
+}
+int add_subcategory(int parent_category, const SCP_string& name)
+{
+	auto& global = globals();
 
-	int subcategory_id = (max_subcategory_value + 1) | parent_category;
+	int subcategory_id = global.next_free_subcategory_id++;
 	op_submenu.push_back({ name, subcategory_id });
+	global.subcategory_to_category[subcategory_id] = parent_category;
```

Comment:
The cleaner way to do this is to do an emplace(subcategory_id, parent_category) here.